Add support for secp256k1 curve

erskingardner opened the pull request:
This adds support for the secp256k1 curve. 

The curve now has [an official IANA HPKE KEM identifier](https://www.iana.org/assignments/hpke/hpke.xhtml).

 p.s. I'm pretty new to Rust so any feedback is very welcome!

erskingardner commented on rust_crypto_provider/src/lib.rs:
```diff
@@ -103,7 +119,11 @@ impl HpkeCrypto for HpkeRustCrypto {
             KemAlgorithm::DhKemP256 => {
-                let sk = SecretKey::from_slice(sk).map_err(|_| Error::KemInvalidSecretKey)?;
+                let sk = p256SecretKey::from_slice(sk).map_err(|_| Error::KemInvalidSecretKey)?;
+                Ok(sk.public_key().to_encoded_point(false).as_bytes().into())
```
I believe this `to_encoded_point` method is the same whether it's imported from `p256` or `k256` but I wanted to call it out to double check.

erskingardner commented:
@franziskuskiefer anything I can do here to help get this PR reviewed?

franziskuskiefer reviewed (changes requested):
Thanks for the contribution! (I just had to find the time to look at it ...)
It generally looks good to me. Only a few nits.

erskingardner commented on tests/test_hpke.rs:
Testing like this with vectors is new for me. Are there any resources you can point me towards so I can better understand how to generate these test vectors?

erskingardner replied:
ahhh.. I think I found them... https://www.rfc-editor.org/rfc/rfc9180.html#name-dhkemp-256-hkdf-sha256-hkdf-s

erskingardner replied:
Found them: https://www.ietf.org/archive/id/draft-wahby-cfrg-hpke-kem-secp256k1-01.html#appendix-B

I'll add them now.

erskingardner commented on CHANGELOG.md:
```diff
@@ -5,6 +5,10 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [0.2.1] - 2024-07-18
```
@franziskuskiefer let me know if you'd rather have this under the 0.2.0 version or if you're planning on cutting a new release.

erskingardner commented on tests/test_vectors.json:
```diff
@@ -153310,5 +153310,203 @@
-    }
+    },
+    {
+        "mode": 0,
```
I've added the `base` mode test vectors here. 

They came from [this doc](https://www.ietf.org/archive/id/draft-wahby-cfrg-hpke-kem-secp256k1-01.html#name-test-vectors). 

Given my lack of experience with vectors, I wanted to point out a few things:

1. I had to add a few fields to the vectors from that doc to make them match the format and work here.
    1. The `enc` field is the same as the `pkEm` field. 
    2. The `key_schedule_context` and `secret` fields are blank.
    3. The `encryption` and `exports` fields are also blank.

erskingardner commented on tests/k256_auth.json:
```diff
@@ -0,0 +1,27 @@
+[
+  {
```
I couldn't get the `auth` mode vectors to work. I was getting failing tests. I'm not sure if it's because the format of the vectors is slightly different or if I actually have an error in the implementation somewhere. I checked in this single auth vector to see if anything jumped out at you that might be wrong with it.

franziskuskiefer replied:
`derive_key_pair` appears to generate the wrong keys based on `ikmS`. That's not too much code but something seems wrong in there. Or might also be that the test vector is wrong. When you checked your code here you could also check the code that gnereated this to make sure it's not doing something wrong.

franziskuskiefer commented on tests/k256_auth.json:
```diff
@@ -0,0 +1,27 @@
+    "skRm": "040fc95447fff5a811321da69ff4655d185d58edef93453ab23dfb1be2f02702",
+    "pkRm": "04a3935d9f2ea9c4b23cdf49f4761625b2acbc1fc89532fe2c3af9d1b1c61b9f167f61ba6125d47151df26e2ecfa851bd79719c99ff354c9b9e9619f25cb6ba6d7",
+    "ikmS": "9fbf7fcf111cc65b6079290c65d0839396104f2dfd39ad34196a4b29d4122383",
+    "skSm": "040fc95447fff5a811321da69ff4655d185d58edef93453ab23dfb1be2f02702",
```
I'm a little suspicious of these test vectors. The value here is the same as for `skRm` but with a differet `ikmR` value.

erskingardner replied:
Yeah, I noticed that too. They don't really seem complete. I actually emailed with the author of that update proposal so I'll drop him an email and ask about the vectors.

franziskuskiefer replied:
Any news on this? I'm happy to get this in when we figured out what's up with the test vectors and can agree on a set that works.

erskingardner replied:
Unfortunately no update. I've been on holiday the last few days so haven't been working on it. I've emailed the original author of the IANA update (who submitted the test vectors) but I've not gotten a response on that.

franziskuskiefer replied:
I've been out as well... How about generating a few new values with their implementation to see if they produce the same results? Maybe only the test vectors are broken. I'm fine using different values. I just want to make sure that the implementations agree.

erskingardner replied:
I've been in touch with the other dev who is also going to have a look at the values too. I'm flying today and tomorrow but I'll be looking into this properly next week.

erskingardner replied:
Hey I've spent the morning on this and from what I can tell, we have the `base` values all correct. (without any encryption or export checks). The `auth` values are don't have a set of sender keys. 

Unrelated to the current issues, why are the private keys and `enc` values are 64 byte strings instead of 32 byte strings in the test vectors?

erskingardner replied:
nevermind that last question. monday morning 🤦

erskingardner commented:
@franziskuskiefer I've put all the k256 vectors in a single file and added a few of the auth vectors (while changing the init values so we get passing tests there). For now I've left the kat tests pointing to this k256.json file, but I'll merge these to the main test_vectors.json file when we're ready to merge. 

A few questions for you: 

1. Do we think testing just the key generation and values of the other context fields is enough or do we need to have `encryptions` and `exports` for each of the k256 tests as well? 
2. This probably depends on the answer to 1 but, right now the sender and receiver keys in the auth tests are the same. do you think we should change those to be different?

franziskuskiefer commented:
> 1. Do we think testing just the key generation and values of the other context fields is enough or do we need to have `encryptions` and `exports` for each of the k256 tests as well?

I think it would be nice to have everything covered by tests. Since we have the code already, this shouldn't be too much work to run for k256 as well I'd hope?

> 2. This probably depends on the answer to 1 but, right now the sender and receiver keys in the auth tests are the same. do you think we should change those to be different?

It shouldn't make a difference really, but it looks a little odd of course. If it's not a big hassle, having different keys would be nice.

erskingardner commented:
@franziskuskiefer I've just pushed an update with new test vectors (courtesy of DanGould) that include encryptions and exports. I'm still getting some odd behavior with them though. 

1. The encryptions and exports all pass the tests
2. The `direct_ctx` check against the `key`, `nonce`, and `exporter_secret` are failing still. 

I'd really appreciate if you could have a look at that and let me know if you think that's an issue with my changes to add k256 or if potentially it's an issue with the test vectors.

franziskuskiefer commented:
> 2\. The `direct_ctx` check against the `key`, `nonce`, and `exporter_secret` are failing still. (lines 155-157)

Hm, when you enable the `hpke-test-prng` feature, you see those tests fail as well. It looks like the code doesn't agree with the test vectors on the `key`. Looking at some intermediate values it looks like the `shared_secret` doesn't seem to match.

Can you post a run where all the intermediate values are printed (just like the test vector)? Then we should see where the discrepancy is and whether the code here or the test vector is wrong.

DanGould commented:
You're right @franziskuskiefer, the test vectors were not right. I generated new ones which work on both my rust-bitcoin k256 implementation as well as one depending on the C version of libsecp256k1 https://github.com/payjoin/bitcoin-hpke/blob/01863441345e244de8f58139371671125da7d496/test-vectors-k256.json

DanGould commented:
@franziskuskiefer'

You're right @franziskuskiefer, the test vectors were not right. I generated new ones which seem to now pass on all 3 implementations I have available: the one here, my `rust-hpke/k256` version depending on RustCrypto k256, and my `rust-hpke/secp` version depending on [rust bindings to the C version of libsecp256k1](https://docs.rs/secp256k1/latest/secp256k1/)

I've tested that this commit runs here in `hpke-rs` with the `hpke-test-prng` feature enabled. let me know what you think. If we have well generated and reproducible test vectors, I'd like to update the RFC draft so that others can find them.

https://github.com/cryspen/hpke-rs/commit/e3a8c8ef222f3aee547325aed7236473ff42eec8

franziskuskiefer commented:
Great! @erskingardner can you update the PR and check if all the tests are passing now?

erskingardner commented:
@DanGould you're a legend! thanks! I'll have a look at this first thing tomorrow.

erskingardner commented:
Actually, I couldn't wait! :) I'm getting passing tests all around 🟢

If you run the tests with `RUST_LOG=trace cargo test --test test_hpke_kat` you'll see the trace output go through all the vectors (including the k256 vectors that I added to the main `test_vectors.json` file. 

@franziskuskiefer I think this looks good. 

Super well done @DanGould 👏 So good to have a set of vectors that pass on three separate implementations.

DanGould commented:
@erskingardner thanks for the quick turnaround. note that the original test_vectors.json are reproducible from the [go-hpke repository](https://github.com/cisco/go-hpke) which is why I originally committed it as a separate file. No problem combining them just noting the original rationale.

franziskuskiefer commented on tests/test_hpke_kat.rs:
```diff
@@ -213,10 +218,16 @@ fn kat<Crypto: HpkeCrypto + 'static>(tests: Vec<HpkeTestVector>) {
 
         // Encrypt
-        for (_i, encryption) in test.encryptions.iter().enumerate() {
+        log::trace!(
```
This seems redundant given the earlier println.

franziskuskiefer reviewed (commented):
> note that the original test_vectors.json are reproducible from the [go-hpke repository](https://github.com/cisco/go-hpke) which is why I originally committed it as a separate file.

I agree, let's use two files.

Otherwise this looks good to me. 

The fuzzer is failing, but that should be easy to fix.

erskingardner commented:
Moved those tests back to individual files @franziskuskiefer and adjusted the test file to run both sets of vectors. 

I have no idea how the fuzzer works. Happy to try and fix it if you want to explain the basics of how to run it.

franziskuskiefer commented on fuzz/fuzz_targets/base.rs:
```diff
@@ -8,6 +8,7 @@ fuzz_target!(|data: &[u8]| {
     let mut hpke = Hpke::<hpke_rs_rust_crypto::HpkeRustCrypto>::new(
         HpkeMode::Base,
         KemAlgorithm::DhKemP256,
+        KemAlgorithm::DhKemK256,
```
You can run the fuzzer with `cargo +nightly fuzz run base` and will see the compilation error. This line should go.

erskingardner commented:
fixed the fuzzer.

franziskuskiefer commented:
CI was hanging. I hope it goes through now. Then we can get this merged.

franziskuskiefer commented:
We need to update CI first in #69. Looks like we didn't update macos CI runners in a while.

erskingardner commented:
Nice! 👍

franziskuskiefer commented:
👍🏻  let me know if you need a new release soonish. I plan to do one, but there's no timeline for it yet.

erskingardner commented:
No rush on the release from my side. 

> On 4 Sep 2024, at 08:51, Franziskus Kiefer ***@***.***> wrote:
> 
> 
> 👍🏻 let me know if you need a new release soonish. I plan to do one, but there's no timeline for it yet.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/cryspen/hpke-rs/pull/66#issuecomment-2328054548>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AABRRAHXKPCMCJTMVPUYV73ZU2UYBAVCNFSM6AAAAABKPGZS7CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMRYGA2TINJUHA>.
> You are receiving this because you were mentioned.
>
